src/pin.ts

```diff
@@ -0,0 +1,419 @@
+ * This is ecosystem-agnostic and separate from package.json
+ */
+interface CheloniaConfig {
+  contracts: {
+    [contractName: string]: {
+      version: string
+      path: string // relative path to the manifest file
+    }
+  }
+  // Future: can add other Chelonia-specific configuration here
+  // like deployment targets, signing keys, etc.
+}
+
+/**
+ * Contract file information extracted from manifest
+ */
+interface ContractFiles {
+  main: string
+  slim?: string
+}
+
+/**
+ * ContractPinner class handles pinning individual contracts to specific versions
+ * 1. Uses chelonia.json for Chelonia-specific configuration (separate from Node.js package.json)
+ * 2. Supports independent versioning for each contract (contracts can have different versions)
+ * 3. Organizes contracts in a structured directory layout: contracts/<name>/<version>/<files...>
+ * 4. Performs selective updates - only modifies the specific contract being pinned
+ * 5. Maintains compatibility with chel manifest and chel deploy workflows
+ *
+ * The pinning process copies contract files from manifest to create new versions,
+ * allowing developers to iterate on contracts while maintaining version history.
+ */
+class ContractPinner {
+  private projectRoot: string
+  private options: PinOptions
+  private cheloniaConfig: CheloniaConfig
+
+  constructor (projectRoot: string, options: PinOptions) {
+    this.projectRoot = resolve(projectRoot)
+    this.options = options
+    // Initialize with empty config - will be loaded from chelonia.json if it exists
+    this.cheloniaConfig = { contracts: {} }
+  }
+
+  /**
+   * Pin contracts with version validation - only pins if CLI version matches manifest version
+   * @param version - The version to pin to (must match manifest version)
+   * @param manifestPath - Path to the manifest file (required)
+   */
+  async pin (version: string, manifestPath?: string) {
+    if (!version || typeof version !== 'string') {
+      throw new Error('Usage: chel pin <version> <manifest-file-path>')
+    }
+
+    // If no manifest path provided, show usage and exit
+    if (!manifestPath) {
+      await this.showUsage()
+      return
+    }
+
+    console.log(colors.cyan(`📌 Requesting pin to version: ${version}`))
+    console.log(colors.gray(`Manifest: ${manifestPath}`))
+
+    // Load existing chelonia.json configuration
+    await this.loadCheloniaConfig()
+
+    // Validate that the manifest file exists
+    const fullManifestPath = join(this.projectRoot, manifestPath)
+    if (!existsSync(fullManifestPath)) {
+      throw new Error(`Manifest file not found: ${manifestPath}`)
+    }
+
+    // Parse the manifest to get contract information and version
+    const { contractName, contractFiles, version: manifestVersion } = await this.parseManifest(fullManifestPath)
+    console.log(colors.blue(`Contract name: ${contractName}`))
+    console.log(colors.blue(`Manifest version: ${manifestVersion}`))
```

Comment:
Nit: the name chosen here is ambiguous as, without context, it could be referring to two different things: the manifest file format version or the contract version (i.e., the contract version the manifest claims it is for). Since this seems to refer to the latter and not the former, I'd rename it to `contractVersion`.